C/common/storage_asset_tracking.cpp

```diff
@@ -0,0 +1,274 @@
+
+using namespace std;
+
+StorageAssetTracker *StorageAssetTracker::instance = 0;
+
+/**
+ * Get asset tracker singleton instance for the current south service
+ *
+ * @return	Singleton asset tracker instance
+ */
+StorageAssetTracker *StorageAssetTracker::getStorageAssetTracker()
+{
+	return instance;
+}
+
+/**
+ * Release the storage asset tracker singleton instance 
+ *
+ * @return      void 
+ */
+
+void StorageAssetTracker::releaseStorageAssetTracker()
+{
+        if (instance)
+                delete instance;
+       	instance = nullptr; 
+}
+
+
+/**
+ * AssetTracker class constructor
+ *
+ * @param mgtClient		Management client object for this south service
+ * @param service  		Service name
+ */
+StorageAssetTracker::StorageAssetTracker(ManagementClient *mgtClient, std::string service) 
+	: m_mgtClient(mgtClient), m_service(service), m_event("store")
+{
+	instance = this;
+}
+
+/**
+ * Fetch all storage asset tracking tuples from DB and populate local cache
+ *
+ * Return the vector of deprecated asset names
+ *
+ */
+void StorageAssetTracker::populateStorageAssetTrackingCache()
+{
+
+	try {
+		std::vector<StorageAssetTrackingTuple*>& vec = m_mgtClient->getStorageAssetTrackingTuples(m_service);
+		for (StorageAssetTrackingTuple* & rec : vec)
+		{
+			auto it = storageAssetTrackerTuplesCache.find(rec);
+			if (it == storageAssetTrackerTuplesCache.end())
+			{
+				// tuple not found in cache , so add it
+				storageAssetTrackerTuplesCache.insert(rec);
+			}
+			else
+			{
+				// tuple present and count value < count of reading, update cache
+				if ((*it)->m_maxCount < rec->m_maxCount)
+				{
+					storageAssetTrackerTuplesCache.erase(it);
+					storageAssetTrackerTuplesCache.insert(rec);
+				}
+				else if ((*it)->m_maxCount == rec->m_maxCount)
+				{
+					// case where counts are same but datapoints are different
+					// "a", "b", "c" and "a", "b", "foo"
+					// keep both the records
+					if ((*it)->m_datapoints.compare(rec->m_datapoints))
+					{
+						storageAssetTrackerTuplesCache.insert(rec);
+					}
+				}
+			}
+
+			Logger::getLogger()->debug("%s:%d Added storage asset tracker tuple to cache: '%s'", __FUNCTION__, __LINE__,
+					rec->assetToString().c_str());
+		}
+		delete (&vec);
+	}
+	catch (...)
+	{
+		Logger::getLogger()->error("%s:%d Failed to populate storage asset tracking tuples' cache",  __FUNCTION__, __LINE__);
+		return;
+	}
+
+
+	return;
+}
+
+/**
+ * Find whether the Storage asset tracking tuple exists in the cache or not
+ *
+ * Return the pointer to the tuple 
+ *
+ * @param tuple        StorageAssetTrackingTuple Type
+ * @return	       A pointer to StorageAssetTrackingTuple in cache or null
+ */
+
+
+StorageAssetTrackingTuple* StorageAssetTracker::findStorageAssetTrackingCache(StorageAssetTrackingTuple& tuple)	
+{
+	StorageAssetTrackingTuple *ptr = &tuple;
+	std::unordered_multiset<StorageAssetTrackingTuple*>::const_iterator it = storageAssetTrackerTuplesCache.find(ptr);
```

Comment:
this can return multiple iterators for a given service/asset/plugin/... tuple? If so, handle that.

Reply:
handled that